db/db_impl/db_impl_compaction_flush.cc

```diff
@@ -806,6 +806,25 @@ Status DBImpl::CompactRange(const CompactRangeOptions& options,
                               end_with_ts);
 }
 
+Status DBImpl::IncreaseFullHistoryTsLow(ColumnFamilyData* cfd,
+                                        std::string ts_low) {
+  VersionEdit edit;
+  edit.SetColumnFamily(cfd->GetID());
+  edit.SetFullHistoryTsLow(ts_low);
+
+  InstrumentedMutexLock l(&mutex_);
+  std::string current_ts_low = cfd->GetFullHistoryTsLow();
+  const Comparator* ucmp = cfd->user_comparator();
+  if (!current_ts_low.empty() &&
+      ucmp->CompareTimestamp(ts_low, current_ts_low) < 0) {
+    return Status::InvalidArgument(
+        "Cannot decrease full_history_timestamp_low");
+  }
+
+  return versions_->LogAndApply(cfd, *cfd->GetLatestMutableCFOptions(), &edit,
```

Comment:
There is an edge case here.
Since `LogAndApply()` can release and acquire the dbmutex, multiple concurrent calls to `IncreaseFullHistoryTsLow()` may interleave. It is possible for one thread to attempt to increase the threshold to 5, while another thread increases the threshold to 10. Either thread can precede the other when writing to the MANIFEST. It's possible that `LogAndApply()` returns OK, but the actual ts_low is 10. In this case, it may be better to inform the caller that the actual ts_low is higher so that this caller will not make the wrong assumption such that data with ts between [5, 10) will be preserved.